Specify which exceptions SHOULD be recorded

Oberon00 opened the pull request:
Replaces #521 "Semantic conventions for when to record exceptions.

Issues that came up in this PR but should be handled separately:
* #764 "Exception semantic conventions and Go error return values"

## Change summary

* State that exceptions that leave the scope of a Span SHOULD be recorded.

## Major changes this PR underwent

* This PR originally proposed that exceptions other than those that leave the scope of a span MUST NOT be recorded. This was replaced by adding the `exception.left_scope` attribute, now renamed to `exception.escaped`. Commit: 11f0543; Discussion: https://github.com/open-telemetry/opentelemetry-specification/pull/761#discussion_r466234754
* The `exception.escaped` (aka `exception.left_scope`) attribute was removed from this PR into #784 because it is not as important and more controversial than the rest of this PR. Commit: 80ea27d; Discussion: [SIG meeting \#61](https://docs.google.com/document/d/1-bCYkN-DWJq4jw1ybaDZYYmx-WAe6HnwfWbkm8d57v8/edit#heading=h.fyk5oh15kuih)

Oberon00 commented:
About the [failing EasyCLA check](https://github.com/open-telemetry/opentelemetry-specification/pull/761#issuecomment-669754048), see https://github.com/open-telemetry/opentelemetry-specification/pull/521#issuecomment-669751034

iNikem commented on specification/trace/semantic_conventions/exceptions.md:
```diff
@@ -13,7 +13,12 @@ exceptions.
-An exception SHOULD be recorded as an `Event` on the span during which it occurred.
+An unhandled exception that leaves the scope of a span
+SHOULD be recorded as an `Event` on that span.
+Other (handled, not leaving a span's scope) exceptions MUST NOT be recorded.
```
First, why do you want to forbid application developers to record any exception that they deem necessary?

Second, what about e.g. errors in Go? AFAIU it is incorrect to say that Go function that returned error "ended
 because of stack unwinding".

Oberon00 replied:
About Go: I was hoping for this kind of feedback, because I have no idea about Go. From what I know, I think you could call the `if error return error` pattern "manual stack unwinding". But do Go errors even qualify as exceptions at all? Do they typically have a stack trace or a distinct type? If not, I'm not sure if the exception reporting conventions should be used at all for them, or if a separate convention would be better.

iNikem replied:
We already mention Go in the semantic convention for exceptions: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/exceptions.md#stacktrace-representation.

My opinion is that we should NOT forbid developers to record any exception they like. We should _recommend_ (as in SHOULD) to record "an unhandled exception that leaves the scope of a span"

Oberon00 replied:
@open-telemetry/go-approvers WDYT? Would a separate semantic conventions for Go(-like) error values make sense? Or do you consider them to be logically exceptions?

MrAlias replied:
Yeah, in Go errors are indeed not equivalent to logical exceptions, but I think the way the specification is written it would be interpreted the way intended by a Go programer. There is a common understanding of errors that the caller handels and others that are unrecoverable for and are propagated upstream, as you pointed out.

I have the same question as @iNikem though as to why you would want to forbid application developers from recording any exceptions.  That level of normative specification [are reserved for incompatibility or security issues](https://tools.ietf.org/html/rfc2119#section-6). From what I can tell this is more of a subjective desire to not have a lot of recorded exceptions? Have I missed something?

MrAlias replied:
It might be possible to make this more language agnostic by starting the section defining an exception as an unrecoverable error. Or, instead just using that term instead of exception throughout.

Oberon00 replied:
The intent was to keep exception and generic error reporting separate. Go errors might be sufficiently exception-like though.

MrAlias replied:
Oh, interesting, I might not have understood that correctly then. It sounds like this is more akin to Go's concept of a [panic](https://golang.org/ref/spec#Handling_panics).

Oberon00 replied:
I created #764. Let's move the discussion about Go errors out of this PR.

Oberon00 commented on specification/trace/semantic_conventions/exceptions.md:
```diff
@@ -13,7 +13,12 @@ exceptions.
-An exception SHOULD be recorded as an `Event` on the span during which it occurred.
+An unhandled exception that leaves the scope of a span
+SHOULD be recorded as an `Event` on that span.
+Other (handled, not leaving a span's scope) exceptions MUST NOT be recorded.
```
Starting a separate thread for this part of https://github.com/open-telemetry/opentelemetry-specification/pull/761/files#r466226987:

@iNikem 
> First, why do you want to forbid application developers to record any exception that they deem necessary?

I originally suggested in #747 to allow all kinds of recorded exceptions and providing a way to distinguish them. The enum I came up with there was not well-received, and I was hoping to get some feedback on this PR here what other exceptions might be deemed necessary to be recorded and maybe come up with a better way to distinguish them than the enum I proposed in #747 (starting from the minimal "there is only one kind of useful exception so we need no way to distinguish them" stance).

iNikem replied:
Yes, I am of opinion is that developers can decide if they want to record an exception or not, and they will not need further annotations. And auto-instrumentations cannot discern/predict exception's future so that can only record one type of exception.

So what we probably can do, is to make a recommendation for _auto-instrumentations only_ to record _only_ those exceptional situations that leave span boundaries. Which would cover to error singling mechanism as well. And we should not make any recommendations or put any constraints on manual exception recording.

Oberon00 replied:
> And auto-instrumentations cannot discern/predict exception's future so that can only record one type of exception.

Agreed. I never understood why you thought #747 was a problem in this regard though.

> we should not make any recommendations on manual exception recording.

Disagree 😃 I think non-binding recommendations are fine.

> or put any constraints on manual exception recording

That I can agree with though. And I can update this PR.

Oberon00 replied:
> developers can decide if they want to record an exception or not, and they will not need further annotations.

I can only talk about the Dynatrace backend. There we would very much like to distinguish at least "unhandled" vs "don't know" (if we can have a finer distinction that's even better). So maybe a boolean `exception.unhandled` could be added, which must be set to true if the exception is known to be an "unhandled exception that leaves the scope of a span". I don't think that puts too much of a burden on anyone.

anuraaga replied:
Can you point me at some instrumentation (not otel I presume) that works with the dynatrace backend that records that? It'd help me understand how to implement this spec :D

Oberon00 replied:
Unfortunately, I don't think we have any open source instrumentation at the moment 😕. Currently in Dynatrace we have our own semantic conventions for exceptions which are more or less the same as the OpenTelemetry ones but we don't use events but span attributes, i.e. we can only record a single exception. And in these, we have this paragraph:

> These attributes MUST NOT be used except for exceptions causing the failure of the operation that the span represents. Therefore only one exception chain can be reported at a span and this exception indicates the reason why the span may represent a failed transaction.

(exception chain means Java-like caused-by chains here).

This was something we came up with because we needed something quick (it predates the OTel conventions) and it does not really adhere to the error vs exception separation, because it assumes that instrumentations would only record exceptions that they know would be errors. Basically what I'm trying to do here is to extend the OpenTelemetry semantic conventions to be able to heuristically determine a "probably most relevant" exception in case multiple are recorded, or maybe lower the severity with which the exception is displayed in the UI if an exception was handled.

As for implementation, our Java instrumentation currently uses an utility function `endSpanWithException(Span span, Throwable throwable)` that ends the span and attaches the exception in one operation.

We would be interested in seeing more exceptions too, but if we would just record all exceptions as events, we would muddy the waters here and lose the information which exception was unhandled and which was not.

anuraaga replied:
Got it - I guess one thing that came to mind is that the Java instrumentation currently only ever adds one exception to a span in practice. I guess they're all the unhandled exception. Handled exceptions don't make it to instrumentation by nature of them having been handled so they don't leak back into the framework, which is what's instrumented.

Does that mean that we want a differentiation between instrumentation exceptions (gotten from the framework) and user recorded ones (handled not returned back to the framework, e.g., servlet)?

Oberon00 replied:
I also think that the Java autoinstrumentation would only ever record unhandled exceptions. I think user-recorded exceptions could be both though, so it's not a distinction auto-instrumentation vs user-recorded. You could probably apply the rule "not unhandled => most likely user-recorded" but not the other way round I think.

Oberon00 replied:
Currently I'm thinking of changing the PR to allow any exception, add a SHOULD for auto-autoinstrumentations to handle scope-leaving exceptions and add an `exception.left_scope` or `exception.unhandled` attribute that tells whether the exception left the scope of the span.

Oberon00 replied:
I added `exception.left_scope` in 9947341.

arminru commented on CHANGELOG.md:
```diff
@@ -10,6 +10,9 @@ the release.
 New:
 
 - Add resource semantic conventions for operating systems ([#693](https://github.com/open-telemetry/opentelemetry-specification/pull/693))
+- Add Span API and semantic conventions for unhandled exceptions
```
```suggestion
- Add Span API and semantic conventions for recording exceptions
```

The content of this PR has changed in the meantime.

bogdandrutu commented on specification/trace/semantic_conventions/exceptions.md:
```diff
@@ -13,9 +13,19 @@ exceptions.
-An exception SHOULD be recorded as an `Event` on the span during which it occurred.
+An unhandled exception that leaves the scope of a span
+SHOULD be recorded as an `Event` on that span.
+An exception is considered to leave the scope of a span if the span is ended
+while the exception is still "in flight"
```
How can you determine this? Where is the "happens before" relationship defined and where is it guaranteed, between handling the exception and span end?

Oberon00 replied:
Often instrumentation is implemented by effectively wrapping a method in a try/catch/finally block like this:

```
final Span span = startSpan(...);
try {
originial code
} catch (Throwable e) {
 span.recordException(e, /*left_scope=*/true);
} finally {
 span.end();
}
```

Here you can clearly determine that the span is ended while the exception is still in flight. Of course if the exception is recorded by code separate from the code that controls the span's scope, you cannot determine left_scope, so just don't set it.

This is the main case this spec is designed for. I will add this example to the text.

Oberon00 replied:
For example, the Python API supports this out of the box. You can use a span as a Python context manager and any exception that leaves this scope is recorded (currently only the status code is set). So to implement this semantic conventions, you basically only would insert `, left_scope=true` at https://github.com/open-telemetry/opentelemetry-python/blob/b3c2a0372e27eff4594891db8ad04c438e9fdf4c/opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py#L677 once it was refactored to use record_exception.

Oberon00 replied:
I think the key point here is not to think about the gneneral case "how can I determine whether any exception will be leaving the scope of the current span at some point", but to realize that there are many common cases where, if you call recordException at the right point, it is absolutely trivial to know for sure that the exception will be leaving the scope. In the non-trivial cases, you leave the attribute unset.

Oberon00 replied:
I added the example in c4a40669010588a819036e2927ad78cf7bdb5c14 and one more clarifying note in a55043d553090218b8838bbe4a343f20becaa43b.

Oberon00 commented:
> This PR without clearly stated adds a notion of "back" propagation which we haven't defined yet 

Actually no. In fact I tried very hard to make clear that no back back propagation is needed at all. That's why added

> Note that multiple events (on the same or different Spans) might be logged for the same exception object instance.

and also

> SHOULD be set to true if the exception event **is recorded while observing** the exception leaving the scope of the span.

If you tell me what makes you think of back propagation here, I will try to reword the spec to make it more clear.

> Unclear what do we try to achieve with the special "left_scope" (also name is bad, initially I thought about where is the "right_scope", probably "left_the_scope").

It is a piece of information that can be used to determine which exception is likely to have been a cause for failure. Also it's very easy to implement this in instrumentations (no back propagation needed, you just will have a literal `true` if you implement your instrumentation in certain ways and leave it unset if you implement it in other ways).


> Why do we not say that exception should be recorded in every Span that "observes" it.

But we do?

> An unhandled exception that leaves the scope of a span SHOULD be recorded as an `Event` on that span.

(actually the "unhandled" is redundant here, I will remove it).

We have no SHOULD requirement for a span that catches an exception which occurs somewhere within it, since such exceptions typically go unobserved (you would have to instrument every catch block, which would generate noise for catch blocks that only do some cleanup and then rethrow).

Oberon00 commented:
> also name is bad, initially I thought about where is the "right_scope"

Hmm, what about exit_by_exception?

bogdandrutu commented:
Let's rephrase my concern as:
What do we miss if we don't add the new attribute? Can we add it later? Is it critical right now?

I do understand that you try to not add back propagation but somehow I got that feeling.

Oberon00 commented:
> What do we miss if we don't add the new attribute?

The ability to distinguish exceptions that left the scope of a span (thus being, in a sense, the result/likely "failure" cause of an operation) from those who aren't. I think this is especially useful when multiple exception events are on a span, to pick out the "most important" exception.

> Can we add it later?

I think we can. The API examples show the attribute with a default of `null`.

> Is it critical right now?

I don't think so. But apparently unexpectedly controversial. 😃
The only thing that is critical before the next spec release is tagged is the change-log entry, and based on how this idea has been received so far I'd better break it out in a separate PR.

bogdandrutu commented:
@Oberon00 I am trying to not add any new feature that can be added later and is not critical. Trying to minimize the scope of GA. Happy to debate this feature more after we stabilize the trace part

Oberon00 commented:
I won't be pushing anyone in particular to merge this PR, but I think the language SIG's additional work for this PR would be very small (trivial even if they did not implement recordException yet). So if enough reviewers are interested, I think it's OK to merge this. I added the `release:after-ga` label anyway, since it's not critical.

I could also remove the flag from this PR, leaving only the SHOULD requirement for when to record exceptions (and the example), which would be no implementation work at all (I guess most instrumentations already behave this way). But I already changed this PR one time and it already replaced a previous draft PR, so I don't want to add more confusion here, if not necessary.

Oberon00 commented:
>> also name is bad, initially I thought about where is the "right_scope"
> Hmm, what about exit_by_exception, exit_by_this or exit_scope?

I renamed it to `exception.escaped`. I was not very satisfied with `left_scope` myself tbh.

Oberon00 commented:
As discussed in the SIG meeting today, removed the new semantic attribute (will submit it as a separate PR) and changed after-ga to required-for-ga for the remainder.

CC @bogdandrutu

bogdandrutu commented on specification/trace/semantic_conventions/exceptions.md:
```diff
@@ -13,9 +13,43 @@ exceptions.
 
-An exception SHOULD be recorded as an `Event` on the span during which it occurred.
+An exception that escapes the scope of a span
+SHOULD be recorded as an `Event` on that span.
```
nit: kind of strange line size for this :)

Oberon00 replied:
I rewrapped this line in bd20a1a.

bogdandrutu commented on specification/trace/semantic_conventions/exceptions.md:
```diff
@@ -13,9 +13,43 @@ exceptions.
 ## Recording an Exception
 
-An exception SHOULD be recorded as an `Event` on the span during which it occurred.
+An exception that escapes the scope of a span
+SHOULD be recorded as an `Event` on that span.
```
Does this mean that the other exceptions should not be recorded?

Oberon00 replied:
There is neither no SHOULD NOT and no SHOULD requirement. I can add:

> Other exceptions MAY also be recorded.

bogdandrutu commented on specification/trace/semantic_conventions/exceptions.md:
```diff
@@ -13,9 +13,43 @@ exceptions.
+
 The name of the event MUST be `"exception"`.
 
+Note that multiple events (on the same or different Spans)
+might be logged for the same exception object instance.
```
nit: same feels the line limit size is kind of random compared with other text.

Oberon00 replied:
I usually break at locations that I feel make sense semantically.

Oberon00 replied:
So if the line exceeds 80 characters, I break at a place such that I think the resulting lines would be most useful to comment on on their own.

Oberon00 replied:
I rewrapped this line in bd20a1a.

bogdandrutu commented on specification/trace/semantic_conventions/exceptions.md:
```diff
@@ -13,9 +13,43 @@ exceptions.
 
+Note that multiple events (on the same or different Spans)
+might be logged for the same exception object instance.
+E.g. one event might be logged in an instrumented exception constructor
+and another event might be logged when an exception leaves the scope of a span.
```
Should we do this? Can we avoid it?

Oberon00 replied:
Hmm, it does add information about the path of the exception although the full stacktrace would not be relevant each time. I don't think this can be easily avoided.

bogdandrutu replied:
Hmm.. is the stacktrace not captured when the exception is generated (at least in some languages)? Also what is the point of having stacktrace twice?

Oberon00 replied:
There is no point, but you usually dont't know if the exception was already/will again be recorded.

Maybe a clever exporter can do de-duplication.

Oberon00 replied:
Note that this is a "Note" stating what I think can happen in practice.

To allow improving this situation, we could do something like assign an ID to each exception when it is first observed and only record the ID for each further occurrence. But that's probably an after-ga thing.

tigrannajaryan commented:
@Oberon00 it is not entirely clear who these recommendations are targeted to. Is it for OpenTelemetry SDK developers or for application developers who use the OpenTelemetry in their apps? Can you please clarify the intent?

Oberon00 commented:
At this point, recommendation in semantic conventions are always directed "to whom it may concern".

In languages that only support recordException APIs, this will be something that has to be done by the user (whoever calls these APIs), although SDK implementors should probably mention this spec part in their documentation for this method.

In other languages, where the API can provide more convenient support for this (for example Python already has a context manager for Spans that automatically sets the status code if an exception is thrown, see https://github.com/open-telemetry/opentelemetry-python/blob/b3c2a0372e27eff4594891db8ad04c438e9fdf4c/opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py#L677), the SDK+API may also help implement this SHOULD requirement.

Oberon00 commented:
Closing this, because there does not seem to be enough interest.
